src/Microsoft.AspNet.Routing/BuilderExtensions.cs

```diff
@@ -29,7 +31,51 @@ public static IApplicationBuilder UseRouter(this IApplicationBuilder builder, IR
                 throw new ArgumentNullException(nameof(router));
             }
 
+            if (builder.ApplicationServices.GetService(typeof(RoutingMarkerService)) == null)
+            {
+                throw new InvalidOperationException(Resources.FormatUnableToFindServices(
+                    nameof(IServiceCollection),
+                    nameof(RoutingServiceCollectionExtensions.AddRouting),
+                    "ConfigureServices(...)"));
+            }
+
             return builder.UseMiddleware<RouterMiddleware>(router);
         }
+
+        /// <summary>
+        /// Creates an <see cref="IRouteBuilder"/> with services from the current
+        /// <paramref name="builder.ApplicationServices"/>.
+        /// </summary>
+        /// <param name="builder">The <see cref="IApplicationBuilder"/> to use for resolving services.</param>
+        /// <returns>An <see cref="IRouteBuilder"/> that can be used to specify the routes in an application.</returns>
+        public static IRouteBuilder UseRouteBuilder(this IApplicationBuilder builder)
+        {
+            return builder.UseRouteBuilder(defaultHandler: null);
+        }
+
+        /// <summary>
+        /// Creates an <see cref="IRouteBuilder"/> with services from the current
+        /// <paramref name="builder.ApplicationServices"/>.
+        /// </summary>
+        /// <param name="builder">The <see cref="IApplicationBuilder"/> to use for resolving services.</param>
+        /// <param name="defaultHandler">The default <see cref="IRouter"/> to use for routing requests.</param>
+        /// <returns>An <see cref="IRouteBuilder"/> that can be used to specify the routes in an application.</returns>
+        public static IRouteBuilder UseRouteBuilder(this IApplicationBuilder builder, IRouter defaultHandler)
+        {
+            if (builder == null)
+            {
+                throw new ArgumentNullException(nameof(builder));
+            }
+
+            if (builder.ApplicationServices.GetService(typeof(RoutingMarkerService)) == null)
+            {
+                throw new InvalidOperationException(Resources.FormatUnableToFindServices(
+                    nameof(IServiceCollection),
+                    nameof(RoutingServiceCollectionExtensions.AddRouting),
+                    "ConfigureServices(...)"));
+            }
+
+            return new RouteBuilder(builder, defaultHandler);
```

Comment:
We should also add the middleware to the pipeline here. This is a `Use` method after all.